src/text_analysis_source_impl.rs

```diff
@@ -0,0 +1,203 @@
+//! we can convey data to the `FontFallback::map_characters` method.
+
+#![allow(non_snake_case)]
+
+use std::borrow::Cow;
+use std::cell::UnsafeCell;
+use std::mem;
+use std::ptr::null;
+use std::sync::atomic::AtomicUsize;
+use winapi::ctypes::wchar_t;
+use winapi::shared::basetsd::UINT32;
+use winapi::shared::guiddef::REFIID;
+use winapi::shared::minwindef::{BOOL, FALSE, TRUE, ULONG};
+use winapi::shared::winerror::{E_INVALIDARG, S_OK};
+use winapi::um::dwrite::{DWRITE_NUMBER_SUBSTITUTION_METHOD, DWRITE_READING_DIRECTION,
+    IDWriteNumberSubstitution, IDWriteTextAnalysisSource, IDWriteTextAnalysisSourceVtbl,
+};
+use winapi::um::unknwnbase::{IUnknown, IUnknownVtbl};
+use winapi::um::winnt::HRESULT;
+
+use helpers::ToWide;
+use super::DWriteFactory;
+use com_helpers::{Com, UuidOfIUnknown};
+use comptr::ComPtr;
+
+/// The Rust side of a custom text analysis source implementation.
+pub trait TextAnalysisSourceImpl {
+    /// Determine the locale for a range of text.
+    ///
+    /// Return locale and length of text (in utf-16 code units) for which the
+    /// locale is valid.
+    fn get_locale_name<'a>(&'a self, text_position: u32) -> (Cow<'a, str>, u32);
+
+    /// Get the text direction for the paragraph.
+    fn get_paragraph_reading_direction(&self) -> DWRITE_READING_DIRECTION;
+}
+
+pub struct CustomTextAnalysisSourceImpl {
+    refcount: AtomicUsize,
+    inner: Box<dyn TextAnalysisSourceImpl>,
+    text: Vec<wchar_t>,
+    number_subst: NumberSubstitution,
+    locale_buf: Vec<wchar_t>,
+}
+
+/// A wrapped version of an `IDWriteNumberSubstitution` object.
+pub struct NumberSubstitution {
+    native: UnsafeCell<ComPtr<IDWriteNumberSubstitution>>,
+}
+
+// TODO: implement Clone, for convenience and efficiency?
+
+DEFINE_GUID! {
+    DWRITE_TEXT_ANALYSIS_SOURCE_UUID,
+    0x12345678, 0x1234, 0x5678, 0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc, 0xde, 0xf0
+}
+
+static TEXT_ANALYSIS_SOURCE_VTBL: IDWriteTextAnalysisSourceVtbl =
+        IDWriteTextAnalysisSourceVtbl {
+    parent: implement_iunknown!(static IDWriteTextAnalysisSource,
+                                DWRITE_TEXT_ANALYSIS_SOURCE_UUID,
+                                CustomTextAnalysisSourceImpl),
+    GetLocaleName: CustomTextAnalysisSourceImpl_GetLocaleName,
+    GetNumberSubstitution: CustomTextAnalysisSourceImpl_GetNumberSubstitution,
+    GetParagraphReadingDirection: CustomTextAnalysisSourceImpl_GetParagraphReadingDirection,
+    GetTextAtPosition: CustomTextAnalysisSourceImpl_GetTextAtPosition,
+    GetTextBeforePosition: CustomTextAnalysisSourceImpl_GetTextBeforePosition,
+};
+
+impl CustomTextAnalysisSourceImpl {
+    /// Create a new custom TextAnalysisSource for the given text and a trait
+    /// implementation.
+    ///
+    /// Note: this method only supports a single `NumberSubstitution` for the
+    /// entire string.
+    pub fn new_native(inner: Box<dyn TextAnalysisSourceImpl>, text: Vec<wchar_t>,
```

Comment:
I don't love only supporting a part of the interface here, but I understand doing this for expedience' sake. My only inclination in this case would be to name the methods in such a way as to allow exposing the API in its full generality in the future. For example, we could have something like `from_text_and_number_subst`, reserving the name `new` for the full API in the future. What do you think?

Reply:
Sure.